src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -267,23 +275,122 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+        CloudRegistrationInfo cloudRegInfo)
+        throws CloudRegistrationAuthorizationException, MalformedCloudRegistrationException {
+
+        if (!this.enabled) {
+            throw new UnsupportedOperationException(
+                "cloud registration is not enabled on this Candlepin instance");
+        }
+
+        if (principal == null) {
+            throw new IllegalArgumentException("principal is null");
+        }
+
+        if (cloudRegInfo == null) {
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isEmpty() || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isEmpty() ||
+            cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isEmpty() || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerExistsAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isEmpty() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerExistsAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
+            }
+        }
+
+        String anonymousConsumerUuid = null;
+        if (ownerKey == null || ownerKey.isEmpty() || !result.isEntitled() || !ownerExistsAndPoolExists) {
+
+            // TODO: Schedule anonymous org creation job
+
+            AnonymousCloudConsumer anonConsumer = anonymousCloudConsumerCurator
+                .create(new AnonymousCloudConsumer(cloudAccountId, productId, cloudProviderShortName));
+            anonymousConsumerUuid = anonConsumer.getUuid();
+        }
+
+        String token = anonymousConsumerUuid == null ?
+            this.buildRegistrationToken(principal, ownerKey, TOKEN_TYPE) :
+            this.buildRegistrationToken(principal, cloudAccountId, ANON_TOKEN_TYPE);
```

Comment:
Can we abstain from adding the cloudAccountId here, and leaving the audience field empty? The client has no need for it, so the less data we expose the better.

Reply:
That makes sense, I updated to use an empty string for the audience field